deepchem/models/torch_models/ferminet.py

```diff
@@ -2,27 +2,268 @@
+        for k in range(nucleon_pos.size()[0]):
+          self.pi_down.append(
+              nn.parameter.Parameter(nn.init.normal(torch.Tensor(1))))
+          self.sigma_down.append(
+              nn.parameter.Parameter(nn.init.normal(torch.Tensor(1))))
+
+  def forward(
+      self,
+      molecule: torch.Tensor,
+  ):
+    """
+    Parameters:
+    -----------
+    one_electron_up: np.ndarray
+      Numpy array containing up-spin electron's one-electron feature
+    one_electron_down: np.ndarray
+      Numpy array containing down-spin electron's one-electron feature
+    two_electron_up: np.ndarray
+      Numpy array containing up-spin electron's two-electron feature
+    two_electron_down: np.ndarray
+      Numpy array containing down-spin electron's two-electron feature
+    """
+    one_electron_vector = molecule - self.nucleon_pos
+
+    shape = molecule.size()
+    two_electron_vector = molecule.reshape([shape[0], 1, shape[1], 3
+                                           ]) - molecule
+
+    one_electron_vector = one_electron_vector[0, :, :, :]
+    two_electron_vector = two_electron_vector[0, :, :, :]
+
+    self.one_electron_distance: torch.Tensor = torch.linalg.norm(
+        one_electron_vector, axis=-1)
+    self.two_electron_distance: torch.Tensor = torch.linalg.norm(
+        two_electron_vector, axis=-1)
+
+    # concatenating distance and vectors arrays
+    one_shape = self.one_electron_distance.size()
+    one_distance = self.one_electron_distance.reshape(one_shape[0],
+                                                      one_shape[1], 1)
+    one_electron = torch.cat(
+        (one_electron_vector, one_distance.repeat(1, 1, 3)), dim=-1)
+    one_electron = one_electron[:, :, 0:4]
+
+    two_shape = self.two_electron_distance.size()
+    two_distance = self.two_electron_distance.reshape(two_shape[0],
+                                                      two_shape[1], 1)
+    two_electron = torch.cat(
+        (two_electron_vector, two_distance.repeat(1, 1, 3)), dim=-1)
+    two_electron = two_electron[:, :, 0:4]
+
+    # TODO: Look into batchwise feed of input
+    one_up = one_electron[:, :self.spin[0], :]
+    one_down = one_electron[:, self.spin[0]:, :]
+
+    one_up = one_up.to(torch.float32)
+    one_down = one_down.to(torch.float32)
+    one_electron = one_electron.to(torch.float32)
+    two_electron = two_electron.to(torch.float32)
+    for i in range(0, 2 * self.layers, 2):
+      g_one_up = torch.mean(one_up, dim=0).flatten()
+      g_one_down = torch.mean(one_down, dim=0).flatten()
+      tmp_one_electron = torch.Tensor([])
+      tmp_two_electron = torch.Tensor([])
+      # spin-up electrons
+      for j in range(0, self.total_electron):
+        one_stream = one_electron[j].flatten()
+        two_stream = two_electron[j]
+        g_two_up = torch.mean(two_stream[:self.spin[0]], dim=0).flatten()
+        g_two_down = torch.mean(two_stream[self.spin[0]:], dim=0).flatten()
```

Comment:
isn't self.spin[0] for up electrons? Is this correct for g_two_down?

Reply:
the two_stream tensor consists of both up-spin and down-spin, so we are going to divide it here for calculating g_two_up and g_two_down. So, using ":self.spin[0]" ensures the index 0 to self.spin[0] will be in up-spin and "self.spin[0]:" will ensure the remaining electrons which are down-spin is taken into account